torch/export/dynamic_shapes.py

```diff
@@ -896,3 +903,182 @@ def assoc_shape(t, dynamic_shape):
+            UserErrorType.INVALID_INPUT,
+            "Suggested fixes not found in error message given to parse_suggested_fixes()",
+        )
+
+    # build suggested_fixes dictionary
+    suggested_fixes = {}
+    for fix in suggested_fixes_msg.split("\n"):
+        fix = fix.strip()
+        if match := re.match(r"(.*) = Dim\('(.*)'.*\)", fix):
+            name = match.group(1)
+            _min, _max = None, None
+            if match_min := re.match(r".* = Dim\('.*', min\=([0-9]+).*\)", fix):
+                _min = int(match_min.group(1))
+            if match_max := re.match(r".* = Dim\('.*'.*max\=([0-9]+)\)", fix):
+                _max = int(match_max.group(1))
+            suggested_fixes[name] = Dim(name, min=_min, max=_max)
+        else:
+            name, expr = fix.split(" = ")
+            expr = sympy.sympify(expr)
+            if isinstance(expr, sympy.Number):
+                suggested_fixes[name] = int(expr)  # static, integer
+            else:
+                suggested_fixes[name] = str(expr)  # relation or derived dim
+
+    name_to_dim = get_dim_name_mapping(dynamic_shapes)
+
+    # track derived dim roots
+    roots: Set[str] = set()
+    for k, c in suggested_fixes.items():
+        if not isinstance(
+            c, (int, _Dim, _DerivedDim, str)
+        ):  # check derived dim expression
+            raise UserError(
+                UserErrorType.INVALID_INPUT,
+                (
+                    "Dictionary values to refine_with_suggested_fixes() must be integers, Dims, DerivedDims, "
+                    f"or strings representing derived dims, but instead found {c} of type {type(c)}"
+                ),
+            )
+        if isinstance(c, int) and not (
+            c >= 0 and c <= sys.maxsize - 1
+        ):  # check integer range
+            raise UserError(
+                UserErrorType.INVALID_INPUT,
+                (
+                    f"Integer values to refine_with_suggested_fixes() must be in the range [0, {sys.maxsize - 1}], "
+                    f"but instead found {c}"
+                ),
+            )
+        if isinstance(c, str):  # check dim/derived dim expression
+            expr = sympy.sympify(c)
+            if not _is_supported_equivalence(expr):
+                raise UserError(
+                    UserErrorType.INVALID_INPUT,
+                    (
+                        "String inputs to refine_with_suggested_fixes() must be supported derived dim expressions "
+                        f"of the form a * x + b, but instead found {expr}"
+                    ),
+                )
+            suggested_fixes[k] = expr  # replace strings with sympy.Exprs
+            roots.add(str(list(expr.free_symbols)[0]))
+        if isinstance(c, _DerivedDim):
+            roots.add(c.root.__name__)  # type: ignore[arg-type]
+
+    # check keys are existing dims or new roots
+    for k, c in suggested_fixes.items():
+        if not (k in name_to_dim or k in roots):
```

Comment:
dumb question, what's the difference between `name_to_dim` and `roots`? it also seems like `roots` isn't used anywhere after this

Reply:
Oh not dumb.

name_to_dim tracks the original dynamic shapes, so if key is in there, it's a refinement of an original shape.

roots also contains introduced roots (sometimes we suggest adding a new dim as a root), so then it's a new dim.

If the key is in neither, then we don't know what to do with it.

I'll rewrite this to be less user facing probably.